### PR TITLE
Make the utility kernels nice and vectorized

### PR DESCRIPTION
**Pull request by krzysz00**
Now, utility kernels give each thread a contiguous range of items to
work on and can use miopen.buffer_store to make their operations
vectorized. This also removes explicit bounds checks - the buffer ops
handle those now.

### CODE REVIEW

**Comment by krzysz00** [comment]
On that note, new ISA from PR test #7

```asm
miopen_conv2d_bwd_weight_gkyxc_nhwgc_nhwgk_0:
        s_load_dwordx2 s[0:1], s[4:5], 0x8
        v_lshlrev_b32_e32 v0, 11, v0
        s_mov_b32 s3, 0x27000
        s_mov_b32 s2, 0x90000
        v_lshl_add_u32 v0, s6, 17, v0
        s_mov_b32 s8, -4
        s_mov_b32 s4, 0
.LBB0_1:
        s_mov_b32 s5, s4
        s_mov_b32 s6, s4
        s_mov_b32 s7, s4
        v_mov_b32_e32 v1, s4
        s_add_i32 s8, s8, 64
        v_mov_b32_e32 v2, s5
        v_mov_b32_e32 v3, s6
        v_mov_b32_e32 v4, s7
        s_cmpk_lt_u32 s8, 0x1fc
        s_waitcnt lgkmcnt(0)
        buffer_store_dwordx4 v[1:4], v0, s[0:3], 0 offen
        buffer_store_dwordx4 v[1:4], v0, s[0:3], 0 offen offset:16
        buffer_store_dwordx4 v[1:4], v0, s[0:3], 0 offen offset:32
        buffer_store_dwordx4 v[1:4], v0, s[0:3], 0 offen offset:48
        buffer_store_dwordx4 v[1:4], v0, s[0:3], 0 offen offset:64
        buffer_store_dwordx4 v[1:4], v0, s[0:3], 0 offen offset:80
        buffer_store_dwordx4 v[1:4], v0, s[0:3], 0 offen offset:96
        buffer_store_dwordx4 v[1:4], v0, s[0:3], 0 offen offset:112
        buffer_store_dwordx4 v[1:4], v0, s[0:3], 0 offen offset:128
        buffer_store_dwordx4 v[1:4], v0, s[0:3], 0 offen offset:144
        buffer_store_dwordx4 v[1:4], v0, s[0:3], 0 offen offset:160
        buffer_store_dwordx4 v[1:4], v0, s[0:3], 0 offen offset:176
        buffer_store_dwordx4 v[1:4], v0, s[0:3], 0 offen offset:192
        buffer_store_dwordx4 v[1:4], v0, s[0:3], 0 offen offset:208
        buffer_store_dwordx4 v[1:4], v0, s[0:3], 0 offen offset:224
        buffer_store_dwordx4 v[1:4], v0, s[0:3], 0 offen offset:240
        v_add_u32_e32 v0, 0x100, v0
        s_cbranch_scc1 .LBB0_1
        s_endpgm
```

vs the old ISA
```asm
miopen_conv2d_bwd_weight_gkyxc_nhwgc_nhwgk_0:
        v_lshl_add_u32 v0, s6, 6, v0
        s_mov_b32 s0, 0x24000
        v_cmp_gt_i32_e32 vcc, s0, v0
        s_and_saveexec_b64 s[0:1], vcc
        s_cbranch_execz .LBB0_3
        s_load_dwordx2 s[0:1], s[4:5], 0x8
        v_ashrrev_i32_e32 v1, 31, v0
        v_add_u32_e32 v2, 0xffff8000, v0
        v_lshlrev_b64 v[0:1], 2, v[0:1]
        s_mov_b64 s[2:3], 0
        s_waitcnt lgkmcnt(0)
        v_mov_b32_e32 v3, s1
        v_add_co_u32_e32 v0, vcc, s0, v0
        v_addc_co_u32_e32 v1, vcc, v3, v1, vcc
        v_mov_b32_e32 v3, 0
        s_mov_b32 s4, 0x1bfff
.LBB0_2:
        v_add_u32_e32 v2, 0x8000, v2
        flat_store_dword v[0:1], v3
        v_add_co_u32_e32 v0, vcc, 0x20000, v0
        v_cmp_lt_i32_e64 s[0:1], s4, v2
        s_or_b64 s[2:3], s[0:1], s[2:3]
        v_addc_co_u32_e32 v1, vcc, 0, v1, vcc
        s_andn2_b64 exec, exec, s[2:3]
        s_cbranch_execnz .LBB0_2
.LBB0_3:
        s_endpgm
```

**Review by whchung — APPROVED** [review]
LGTM

**Comment by krzysz00** [comment]
skip-ci due to the very annoying amount of "it works on my machine" we've been dealing with and since this can wait

File: mlir/tools/mlir-miopen-lib/mlir-miopen-lib.cpp [hunk 1]
```diff
@@ -41,6 +42,7 @@ struct MiirHandle_s {
     auto getRegistry = []() {
       DialectRegistry registry;
       registerAllDialects(registry);
+      registerMIOpenDialects(registry);
```
Comment (jerryyin) [inline mlir/tools/mlir-miopen-lib/mlir-miopen-lib.cpp:45]: >  LLVM ERROR: Building op `func.func` but it isn't registered in this MLIRContext: the dialect may not be loaded or this operation isn't registered by the dialect. See also https://mlir.llvm.org/getting_started/Faq/#registered-loaded-dependent-whats-up-with-dialects-management
> ...
> [2022-07-14T18:51:28.912Z] 0.	Program arguments: mlir-rocm-runner --shared-libs=/var/z1_miciadmin/workspace/MLIR_mlir_PR-636/build/external/llvm-project/llvm/lib/libmlir_rocm_runtime.so,/var/z1_miciadmin/workspace/MLIR_mlir_PR-636/build/lib/libconv-validation-wrappers.so,/var/z1_miciadmin/workspace/MLIR_mlir_PR-636/build/external/llvm-project/llvm/lib/libmlir_runner_utils.so --entry-point-result=void

I think according to the stack trace it has less to do with `libMLIRMIOpen` but more to do with `mlir-rocm-runner`.

Maybe try to add a `context.loadDialect<func::FuncDialect>` into `mlir-rocm-runner` to make sure it is loaded.

File: mlir/lib/Dialect/MIOpen/Transforms/AffixTuningParameters.cpp [hunk 2]
```diff
@@ -118,6 +119,20 @@ void AffixTuningParameters::runOnOperation() {
...
+  int64_t numElements = arg.getType().cast<MemRefType>().getNumElements();
+  int64_t blockSize = kUtilityKernelBlockSize;
+  int64_t elemsPerThread = kUtilityKernelElemsPerThread;
+  int64_t gridSize =
+      math_util::integer_divide_ceil(numElements, blockSize * elemsPerThread);
+  for (Operation *op : ArrayRef<Operation *>{convOp, funcOp}) {
+    op->setAttr("grid_size", b.getI32IntegerAttr(gridSize));
+    op->setAttr("block_size", b.getI32IntegerAttr(blockSize));
+    op->setAttr("elems_per_thread", b.getI32IntegerAttr(elemsPerThread));
+  }
```
Comment (jerryyin) [inline mlir/lib/Dialect/MIOpen/Transforms/AffixTuningParameters.cpp]: > [2022-07-14T18:51:21.027Z] 0.	Program arguments: /var/z1_miciadmin/workspace/MLIR_mlir_PR-636/build/bin/miopen-opt -miopen-affix-params -miopen-conv-to-gemm
> ...
>  [2022-07-14T18:51:21.027Z]  #4 0x00007f64c7a86200 mlir::Attribute::getDialect() const /var/z1_miciadmin/workspace/MLIR_mlir_PR-636/build/../external/llvm-project/mlir/include/mlir/IR/Attributes.h:73:0

This feels similar to previous one but in a different component, that funDialect is not loaded in `miopen-opt`

File: mlir/tools/mlir-miopen-lib/mlir-miopen-lib.cpp (hunk 1 above)
Comment (krzysz00) [reply mlir/tools/mlir-miopen-lib/mlir-miopen-lib.cpp:45]: I'd think the error is a consequence of the func not getting lowered to llvm properly

But I wouldn't be seeing the issue if I'd marked these things `kernel`??

Comment (jerryyin) [reply mlir/tools/mlir-miopen-lib/mlir-miopen-lib.cpp:45]: > I'd think the error is a consequence of the func not getting lowered to llvm properly

Agreed, rocm runner shouldn't expect a func op.

> But I wouldn't be seeing the issue if I'd marked these things kernel??

We picked those func in `MIOpen -> GPU` pass that has `kernel` attributes. Sounds like the way to go?

**Comment by krzysz00** [comment]
Today's very fun times
- This PR works on ROCm clang
- It works on GCC 7.5 (what we use in CI) in Debug mode
- It doesn't work with RelWithDebInfo on GCC 7.5

File: mlir/lib/Dialect/MIOpen/Transforms/AffixTuningParameters.cpp [hunk 3]
```diff
@@ -130,7 +130,8 @@ static void setUtilityKernelSizes(OpBuilder &b, Value arg, Operation *convOp,
   int64_t elemsPerThread = kUtilityKernelElemsPerThread;
   int64_t gridSize =
       math_util::integer_divide_ceil(numElements, blockSize * elemsPerThread);
-  for (Operation *op : ArrayRef<Operation *>{convOp, funcOp}) {
+  SmallVector<Operation *, 2> ops = {convOp, funcOp};
+  for (Operation *op : ops) {
```
Comment (jerryyin) [inline mlir/lib/Dialect/MIOpen/Transforms/AffixTuningParameters.cpp:134]: I'm curious how it broke you: Does ArrayRef died after finishing the first loop so effectively the second loop is on garbage? But I also remember the error message does mention about `funcop` being illegal, which means it has looped through the second op successfully. So what actually happens (breaks) that caused those mysterious issues?

Comment (krzysz00) [reply mlir/lib/Dialect/MIOpen/Transforms/AffixTuningParameters.cpp:134]: My gut is telling me it's something about the temporary not living long enough and so getting that chuck of stack space trampled by the function calls in the loop or ... idk, I'd really love Rust's compiler here.

**Pull request merged**